plugins/api/src/replica_close.cpp

```diff
@@ -369,6 +380,16 @@ namespace
 
             const auto is_write_operation = (O_RDONLY != (l1desc.dataObjInp->openFlags & O_ACCMODE));
 
+            // Close the underlying file object.
+            if (const auto ec = close_physical_object(*_comm, l1desc.l3descInx); ec != 0) {
+                log::api::error("Failed to close file object [error_code={}].", ec);
+                if (is_write_operation) {
+                    update_replica_status_on_error(*_comm, l1desc);
+                    irods::experimental::replica_access_table::erase_pid(l1desc.replica_token, getpid());
```

Comment:
I'm not sure that I understand why the replica access table entry was being restored in the old logic. Failing to remove the replica access table entry causes overwrites in S3 to fail because the PID of the new agent is not found in the entry. If the failure occurs in an iCommand, there's no way you will be able to open this object again because the connection closes and the agent goes away.

Would it be appropriate to have the agent teardown logic also call `erase_pid` just in case?

Reply:
Okay, I see. In that case, I think only removing the PID on success makes sense. Once the client disconnects or the agent process is reaped on termination, the PID will be removed from the replica access table if it's not done so already in this function (L453). That way, the connected client can try as many times as needed on failure to close the replica and the PID will be guaranteed to be removed.

> Worst case scenario is the agent factory has to do it (which it does - you'll need to verify what happens if an unknown PID is removed). Does the replica access table throw an exception or ignore it?

I added an assertion to unit_tests/src/test_replica_access_table.cpp which calls `erase_pid` with a PID that does not exist in the replica access table and it did not throw:
```c++
// Erase a PID which does not exist in the replica access table.
REQUIRE_NOTHROW(rat::erase_pid(7000));
```
Is that what you meant by "unknown"?

Reply:
> Is that what you meant by "unknown"?

Yes.

> Okay, I see. In that case, I think only removing the PID on success makes sense.

That sounds reasonable. What happens if the file on disk is closed and then we fail on a catalog update? Will `close_physical_object` succeed on a closed file or are we stuck?

How does this compare to `rsDataObjClose`? Does anything need to change there?
https://github.com/irods/irods/blob/03e7e0515ab3e367c969ddc783b0e0d3dd995924/server/api/src/rsDataObjClose.cpp#L732-L762

Reply:
After some discussion, we have determined that just erasing the PID on success in `replica_close` is probably okay since the tests are passing and there are no anticipated issues with the approach.

If the file on disk is closed and we fail on a catalog update, we will attempt to unlock the data object in all cases (see `update_replica_status_on_error`), in which case the open replica will be stale with the other replica statuses being restored to what they were before they were locked. If this fails, the object is stuck in a locked status because we cannot reach the catalog. In that case, there's nothing we can do. If we want to allow the user to try again, we will need to take a long, hard look at `replica_close` and `rsDataObjClose` (and, by extension, their "open" counterparts). After discussion, we think that's beyond the scope of this change, but should happen at some point.

The logic shown in `rsDataObjClose` there is not going to restore the replica access table entry at all the right times (I think). I've created an issue to address this: https://github.com/irods/irods/issues/6890 I initially thought this code might be dead, but it's not *quite* dead. Again, this is beyond the scope of this change.

I will mark this as resolved and we can raise a new thread if similar concerns arise later.